app/src/main/java/org/mozilla/fenix/utils/BrowsersCache.kt

```diff
@@ -0,0 +1,54 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.utils
+
+import android.content.Context
+import androidx.annotation.VisibleForTesting
+import mozilla.components.support.utils.Browsers
+
+/**
+ * Caches the list of browsers installed on a user's device.
+ *
+ * BrowsersCache caches the list of installed browsers is gathered lazily when it is first accessed
+ * after initial creation or invalidation. For that reason, a context is required every time
+ * the cache is accessed.
+ *
+ * In addition, a context is required every time the cache is accessed because every access
+ * is guaranteed to return cached results only if the context parameter of the access matches the
+ * context parameter given when the cache was filled. This ensures that if the cache is filled with
+ * a list of browsers installed for a particular context, that list will be invalidated and replaced
+ * if the cache is accessed for a different context.
+ *
+ * Users are responsible for invalidating the cache at the appropriate time. It is left up to the
+ * user to determine appropriate policies for maintaining the validity of the cache. As mentioned
+ * above, the cache will be lazily refilled after invalidation. In other words, invalidation is O(1).
+ */
+object BrowsersCache {
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal var cachedBrowsers: Browsers? = null
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal var cachedContext: Context? = null
+
+    @Synchronized
+    fun all(context: Context): Browsers {
+        run {
+            val cachedContext = cachedContext
+            val cachedBrowsers = cachedBrowsers
+            if (cachedContext != null && cachedBrowsers != null && cachedContext == context) {
```

Comment:
I see two issues with this:
1) We're storing an Activity cache statically so this actually creates a memory leak. If we pass in an Application Context instead, the check isn't useful
2) I think this might unnecessarily reset the cache if:
* `IntentReceiverActivity` fetches Browsers
* We fetch Browsers in `MainActivity`

iirc, the problem we were trying to solve was:

> Should there come a time when more than one context is used (not that I can imagine one), we want to make sure that we don't return the wrong result.

Can you describe from a user perspective what this might look like? I'm struggling to come up with one: it might dictate a better solution.

That being said, it sounds like we're trying to solve a problem we don't have yet so I wonder if there's a better way to prevent us from having bugs when that problem comes. For example, we could assume the relationship between `MainActivity` and `IntentReceiverActivity` won't change, for which this class is written, and do something like:
```kotlin
    fun all(context: Context): Browsers {
        require(!(context is MainActivity::class) && !(context is IntentReceiverActivity::class)) {
            "This class is written with assumptions about the lifecycles of MainActivity and IntentReceiverActivity and how that affects cache invalidation. This class needs to be adapted to this new Context."
        }
       // ...
```

Reply:
> We're storing an Activity cache statically so this actually creates a memory leak. If we pass in an Application Context instead, the check isn't useful

I guess we actually empty the cache in `onPause` so the current implementation is fine (assuming `IntentReceiverActivity` has no shenanigans) but it's fragile for future scenarios (e.g. `resetAll` is moved or a new Context is introduced where `resetAll` is used differently).